### DISCUSSION TITLE
How to implement RMSE loss function

### DISCUSSION THREAD

**Discussion by 00houssam00** [MilesCranmer/PySR#553] [Q&A] [answered]
Thank you for this awesome library!!

I have been playing around with PySR for some time. I was trying to create a custom loss function to apply RMSE as the loss but couldn't find how to anywhere in the documentation. is there any way to apply RSME with PySR ?

**Comment by MilesCranmer** [comment, marked as answer]
Thanks! And welcome to the forums :)

Anything which can’t be expressed as a sum of per-row losses you should write a custom loss function for. Here’s RMSE as an example:

```python
loss_function = """
function f(tree, dataset::Dataset{T,L}, options) where {T,L}
    ypred, completed = eval_tree_array(tree, dataset.X, options)
    if !completed
        return L(Inf)
    end
    y = dataset.y
    return sqrt(sum(i -> (ypred[i] - y[i])^2, eachindex(y)))
end
"""

model = PySRRegressor(loss_function=loss_function)
```

**Reply by 00houssam00** [reply]
Amazing, Thank you for the answer :)

**Reply by gm89uk** [reply]
Thank you Miles! Anyway to make this work with batching = True? Couldn't figure it out

**Comment by pukpr** [comment]
Now I really want to experiment with PySR. In the past I've used Eureqa and found it amazing. But more recently I've been using a home-grown optimizer for modeling time-series data. I discovered that a loss function built on a Dynamic Time Warping metric is very helpful to guard against over-fitting, see my blog post https://geoenergymath.com/2024/03/08/dynamic-time-warping.  Now, thinking back to how Eureqa often took a while to close in on a potential solution, a DTW metric as loss function may be very useful in an exploratory mode. It wouldn't be difficult to prototype as a custom loss function with a configurable window size.  See this 2024 paper as well: https://arxiv.org/pdf/2401.10359.pdf

**Reply by MilesCranmer** [reply]
Very cool!! Do you want to make a new discussion thread with this? (i.e., copy this comment to new thread). We can try to get it working there.

**Reply by pukpr** [reply]
Thanks, done >> MilesCranmer/PySR/discussions/563
